### PR TITLE
Added YouTube video downloader

### LINKED ISSUE #602: YouTube Video Downloader with GUI
# Aim

- **YouTube Video Downloader** is an application to download videos from YouTube. This provides users to download videos they need in their devices and watch them offline.
- The objective of this script is to download any type of video in a fast and easy way from YouTube in your device.

# Details

- In this python script, user has to copy the YouTube video URL that they want to download and simply paste that URL in the ‘paste link here’ section and click on the download button, it will start downloading the video. When video downloading finishes, it shows a message ‘downloaded’ popup on the window below the download button.
- It will look like this:

YTVD

### PR DESCRIPTION
**Pull request by mehabhalodiya**
# Description
To implement this, I have used basic concept of python, tkinter and pytube library.

- **Tkinter** is a standard GUI library and it is one of the easiest ways to build a GUI application.
- **pytube** used for downloading videos from youtube

## Have you read the [Contributing Guidelines on Pull Requests](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/CONTRIBUTING.md)?

- [x] Yes
- [ ] No

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] My documentation follows [`Template for README.md`](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/Template%20for%20README.md)
- [x] My changes generate no new warnings
- [x] I have added tests/screenshots that prove my feature is effective or that my feature works.

### CODE REVIEW

File: YouTube-Video-Downloader/youtube_vid_dl.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+root.resizable(0,0)
+root.title("YouTube Video Downloader")
+
+Label(root,text = 'Copy the link of the video you want to download from YouTube', font = 'arial 15 bold').pack()
+
+##enter link
+link = StringVar()
+
+Label(root, text = 'Paste Link Here:', font = 'arial 15 bold').place(x = 270 , y = 60)
+link_enter = Entry(root, width = 80,textvariable = link).place(x = 32, y = 90)
```
Comment (antrikshmisri) [inline YouTube-Video-Downloader/youtube_vid_dl.py:15]: The Entry function doesn't have any return value , why are you assigning it to a variable

Comment (mehabhalodiya) [reply YouTube-Video-Downloader/youtube_vid_dl.py]: Updated!
Thank you :)

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
- Add a progress bar which shows how much the download has been completed.
- Add the option to control the quality of the video (480, 720, ...)

**Comment by mehabhalodiya** [comment]
> * Add a progress bar which shows how much the download has been completed.
> * Add the option to control the quality of the video (480, 720, ...)

I will implement the feature of progress bar.
I am not getting about quality control download!

I need your help! Please help me out, I am ready to work.

**Comment by kaustubhgupta** [comment]
@mehabhalodiya Refer to this: https://python-pytube.readthedocs.io/en/latest/user/streams.html#filtering-streams

As the yt.streams.filter(progressive=True) returns a list of all types of resolutions in a sorted manner, you can modify the script to pick the available qualities and then present the user options to pick
```
[<Stream: itag="18" mime_type="video/mp4" res="360p" fps="30fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" type="video">,
<Stream: itag="22" mime_type="video/mp4" res="720p" fps="30fps" vcodec="avc1.64001F" acodec="mp4a.40.2" progressive="True" type="video">]
```

**Comment by kaustubhgupta** [comment]
@mehabhalodiya updates?

**Comment by kaustubhgupta** [comment]
@mehabhalodiya Final updates: did you found the workaround or should I assign it to someone else?

**Comment by kaustubhgupta** [comment]
@mehabhalodiya If you are unable to add quality options, you can add the progress bar and we will accept that submission

**Comment by mehabhalodiya** [comment]
Okay, I am doing it!

**Comment by kaustubhgupta** [comment]
@mehabhalodiya Updates?

**Comment by mehabhalodiya** [comment]
Now?

```
from tkinter import *
from tkinter.ttk import *
from pytube import YouTube

root = Tk()
root.geometry('700x300')
root.resizable(0,0)
root.title("YouTube Video Downloader")

Label(root,text = 'Copy the link of the video you want to download from YouTube', font = 'arial 15 bold').pack()

# enter link
link = StringVar()

Label(root, text = 'Paste Link Here:', font = 'arial 15 bold').place(x = 270 , y = 60)
Entry(root, width = 80,textvariable = link).place(x = 32, y = 90)

# progress bar widget
progress = Progressbar(root, orient = HORIZONTAL, length = 100, mode = 'determinate')

# function responsible for the updation of the progress bar value
def bar():
    import time
    progress['value'] = 20
    root.update_idletasks()
    time.sleep(1)

    progress['value'] = 40
    root.update_idletasks()
    time.sleep(1)

    progress['value'] = 50
    root.update_idletasks()
    time.sleep(1)

    progress['value'] = 60
    root.update_idletasks()
    time.sleep(1)

    progress['value'] = 80
    root.update_idletasks()
    time.sleep(1)
    progress['value'] = 100

progress.pack(pady=10)

# function to download video
def Downloader():

    url = YouTube(str(link.get()))
    video = url.streams.first()
    bar()
    video.download()
    Label(root, text = 'DOWNLOADED', font = 'arial 15').place(x = 270 , y = 210)

Button(root, text = 'DOWNLOAD', font = 'arial 15 bold', bg = 'white', padx = 2, command = Downloader).place(x = 280 ,y = 150)

root.mainloop()
```

**Comment by kaustubhgupta** [comment]
> Now?
> 
> ```
> from tkinter import *
> from tkinter.ttk import *
> from pytube import YouTube
> 
> root = Tk()
> root.geometry('700x300')
> root.resizable(0,0)
> root.title("YouTube Video Downloader")
> 
> Label(root,text = 'Copy the link of the video you want to download from YouTube', font = 'arial 15 bold').pack()
> 
> # enter link
> link = StringVar()
> 
> Label(root, text = 'Paste Link Here:', font = 'arial 15 bold').place(x = 270 , y = 60)
> Entry(root, width = 80,textvariable = link).place(x = 32, y = 90)
> 
> # progress bar widget
> progress = Progressbar(root, orient = HORIZONTAL, length = 100, mode = 'determinate')
> 
> # function responsible for the updation of the progress bar value
> def bar():
>     import time
>     progress['value'] = 20
>     root.update_idletasks()
>     time.sleep(1)
> 
>     progress['value'] = 40
>     root.update_idletasks()
>     time.sleep(1)
> 
>     progress['value'] = 50
>     root.update_idletasks()
>     time.sleep(1)
> 
>     progress['value'] = 60
>     root.update_idletasks()
>     time.sleep(1)
> 
>     progress['value'] = 80
>     root.update_idletasks()
>     time.sleep(1)
>     progress['value'] = 100
> 
> progress.pack(pady=10)
> 
> # function to download video
> def Downloader():
> 
>     url = YouTube(str(link.get()))
>     video = url.streams.first()
>     bar()
>     video.download()
>     Label(root, text = 'DOWNLOADED', font = 'arial 15').place(x = 270 , y = 210)
> 
> Button(root, text = 'DOWNLOAD', font = 'arial 15 bold', bg = 'white', padx = 2, command = Downloader).place(x = 280 ,y = 150)
> 
> root.mainloop()
> ```

Did you run the script locally?
image

**Comment by mehabhalodiya** [comment]
Yes, me too got stuck in this!
Tell me which lines to be added and improved.

**Comment by kaustubhgupta** [comment]
> Yes, me too got stuck in this!
> Tell me which lines to be added and improved.

as the error suggests, try removing font attributes

**Comment by mehabhalodiya** [comment]
> > Yes, me too got stuck in this!
> > Tell me which lines to be added and improved.
> 
> as the error suggests, try removing font attributes

Yes, but still not working :/

**Comment by kaustubhgupta** [comment]
> > > Yes, me too got stuck in this!
> > > Tell me which lines to be added and improved.
> > 
> > 
> > as the error suggests, try removing font attributes
> 
> Yes, but still not working :/

what's the error now?

**Comment by mehabhalodiya** [comment]
output

**Comment by mehabhalodiya** [comment]
I have removed font, bg and padx attributes.

**Comment by kaustubhgupta** [comment]
> I have removed font, bg and padx attributes.

okay, so now you are getting bar progress with download without any errors? If yes then update the code in PR, I will review and approve asap

**Comment by mehabhalodiya** [comment]
@kaustubhgupta Let me know if I am done with the code so that I can modify!
Give me a sign to move ahead.

**Comment by kaustubhgupta** [comment]
> @kaustubhgupta Let me know if I am done with the code so that I can modify!
> Give me a sign to move ahead.

Um, then copy your code, paste it as gist and share it here, I will run it locally and will let you know the status

**Comment by mehabhalodiya** [comment]
https://gist.github.com/mehabhalodiya/0c4d67f3225ac55a8f7382228b5712b1

**Comment by kaustubhgupta** [comment]
Umm it didn't ran
image

**Comment by mehabhalodiya** [comment]
So now?
I think progress bar messed up all!

**Comment by kaustubhgupta** [comment]
> So now?
> I think progress bar messed up all!

I will re-review all the changes in the morning and will let you know about any changes/

**Comment by kaustubhgupta** [comment]
@mehabhalodiya dropping the idea for the progress bar. It would require creating separate thread for this other than the main GUI thread.

**Review by kaustubhgupta — APPROVED** [review]

**Comment by mehabhalodiya** [comment]
@antrikshmisri @santushtisharma10 Kindly review this.

**Review by santushtisharma10 — APPROVED** [review]
Nice @mehabhalodiya 💯 💯

**Comment by kaustubhgupta** [comment]
@antrikshmisri

**Review by antrikshmisri — APPROVED** [review]

**Review by avinashkranjan — APPROVED** [review]

**Pull request merged**